File: Programmers/Hash/telephone.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool solution(vector<string> phone_book)
{
    sort(phone_book.begin(), phone_book.end());
    vector<string>::iterator it = phone_book.begin();
    while (it != phone_book.end() - 1)
    {
        string prefix = string((*(it + 1)).begin(), (*(it + 1)).begin() + (*it).size());
        cout << prefix << endl;
        if (prefix == *it)
        {
            return false;
        }
        it++;
    }
    return true;
}

int main()
{
    vector<string> phone_book = {"119", "97674223", "1195524421"};
    cout << solution(phone_book) << endl;
    cin.get();
}

File: Programmers/Hash/telephone.py
def solution(phone_book):
    phone_book.sort()
    for i in range(len(phone_book) - 1):
        if len(phone_book[i]) < len(phone_book[i + 1]):
            if phone_book[i + 1][:len(phone_book[i])] == phone_book[i]:
                return False
    return True

def solution1(phone_book):
    phone_book.sort()
    for p, r in zip(phone_book, phone_book[1:]):
        if r.startswith(p):
            return False
    return True

print(solution1(["119", "97674223", "1195524421"]))
print(solution1(["123","456","789"]))
print(solution1(["12","123","1235","567","88"]))